### ISSUE TITLE
Having problems getting [midifile] to work

### ISSUE THREAD

**Issue by dfkettle** [pd-externals/mrpeach#1]
This isn't really an "issue", I don't think, but I'm having problems getting [midifile] to work. I've had no luck creating MIDI files, except for the most trivial of examples. If I try to play back the file, I just get a stream of error messages like the following:

`midifile: No running status on track 0 at 957`

And if I try to play it back using [cyclone/seq], I get the following error messages:

```
seq's warning: missing running status in midi file... skip to end of track
seq's warning: no valid miditracks
```

I've posted this question on the forum as well as the mailing list, but haven't been able to resolve the problem. Is there some place I can find examples, other than what's in the help file for [midifile]? Maybe if I can look at how other people are using it, I can figure out what I'm doing wrong. I've tried Google, but didn't come up with anything.

I'm using the latest version available through Deken, v20200615, (although I don't know if there's a later version that isn't available in Deken yet), under Pd 0.52.1 and Windows 7 (64-bit).

Thanks.

**Comment by danomatika** [comment]
Can you provide a minimum patch which uses `[midifile]` to output a file? If not, then the list of steps you follow for creating a test file for the help patch.

**Comment by dfkettle** [comment]
I'll try, although I haven't been able to boil it down to a minimal patch
yet. Very simple examples work, but what I was trying to do (which is too
complicated to post) doesn't work. It may take me a day or two to come up
with something.

Thanks.

**Comment by dfkettle** [comment]
> Also note that "running status" refers to MIDI allowing the status byte of the message to be left off if it's the same as the previous message: http://midi.teragonaudio.com/tech/midispec/run.htm
> 
> I would double check how you are formatting the MIDI messages being sent into `[midifile]`.

Why is it a warning message? It's OK to omit the status byte, if it's the same as the previous message, isn't it? But [cyclone/seq] is saying the track is invalid. (I'm assuming that I should be able to read MIDI files created by [midifile] using [cyclone/seq] or any other software that reads MIDI files.)

Anyway, I'm not getting that particular error with this example, but I am getting an error message about deleting a temp file (see line 32 in the attached log file), which I don't understand. It seems to create the MIDI file anyway, but doesn't delete the temp file (test.mid0.trk). I don't know why it's saying "Permission denied", I do have full access to that directory.

Also, I've noticed that the number of bytes written to the file is always 1 or 2 less than what's reported in the console log (see line 29). Obviously not a serious issue, but it might cause some confusion.

And finally, [midifile] seems to be ignoring "track" messages. I'm sending it a [track 0( message when writing to the file and a [track 1( message when reading it. But it seems to play the track anyway.

Do you know of any examples from other users that I could look at? There aren't any provided with the external, other than the help file.

[MIDI Sequencer.zip](https://github.com/pd-externals/mrpeach/files/8447227/MIDI.Sequencer.zip)

**Comment by danomatika** [comment]
> Why is it a warning message? It's OK to omit the status byte, if it's the same as the previous message, isn't it?

Yes, but the message indicates it's getting a set of data bytes for a *different message* which cancels the running status. The MIDI spec says running status is valid for voice messages but is automatically cancelled by system messages.

From the link in my previous message:

>System Common Category messages (ie, Status of 0xF0 to 0xF7) cancel any running status. In other words, the message after a System Common message must begin with a Status byte. System Common messages themselves can't be implemented with running status. For example, if a Song Select message was sent immediately after another Song Select, the second message would still need a Status byte.
>
>Running status is only implemented for Voice Category messages (ie, Status is 0x80 to 0xEF).

**Comment by dfkettle** [comment]
All I'm doing is sending lists of three numbers representing Note On messages to [midifile]. I'm not telling [midifile] to omit the status byte (I don't even know how I would do that). If you see something wrong with my patch, let me know, but I'm just doing what it says in the help file, I think.

**Comment by danomatika** [comment]
My previous post was about using the message for debugging the situation, not that I had a solution.

In any case, trying out your patch on my macOS 11.6.1 system with Pd 0.52-2, the latest version of cyclone, and a recent version of [mrpeach/midifile] I compiled about a month ago: your test output file plays back ok in [cyclone/seq] without errors, although it's missing the note offs of course.

Console output:
~~~
midifile verbosity is 3
midifile_write: path = test.mid, fps = 0, tpf = 90
midifile_open_path (relative): /Users/danomatika/Downloads/MIDI/test.mid

midifile: opened /Users/danomatika/Downloads/MIDI/test.mid
midifile_open_track_file: main file path is /Users/danomatika/Downloads/MIDI/test.mid
midifile_open_track_file: track path is /Users/danomatika/Downloads/MIDI/test.mid0.trk
this track (0) is being used for the first time
midifile_open_track_file: main file path is /Users/danomatika/Downloads/MIDI/test.mid
midifile_open_track_file: track path is /Users/danomatika/Downloads/MIDI/test.mid0.trk
midifile_list. j[0]	= 0x90
midifile: j = 0x90
midifile_list. j[1]	= 0x43
midifile: j = 0x43
midifile_list. j[2]	= 0x5A
midifile: j = 0x5A
midifile_list. j[0]	= 0x90
midifile: j = 0x90
midifile_list. j[1]	= 0x40
midifile: j = 0x40
midifile_list. j[2]	= 0x5A
midifile: j = 0x5A
midifile_list. j[0]	= 0x90
midifile: j = 0x90
midifile_list. j[1]	= 0x3C
midifile: j = 0x3C
midifile_list. j[2]	= 0x5A
midifile: j = 0x5A
midifile: wrote 43 to /Users/danomatika/Downloads/MIDI/test.mid
midifile_delete_track_file: main file path is /Users/danomatika/Downloads/MIDI/test.mid
midifile_delete_track_file: "/Users/danomatika/Downloads/MIDI/test.mid0.trk"
midifile_open_path (relative): /Users/danomatika/Downloads/MIDI/test.mid

midifile: opened /Users/danomatika/Downloads/MIDI/test.mid
midifile: Header chunk type: MThd
midifile: Header chunk length: 6
midifile: Header chunk format: 0 (Single multichannel track)
midifile: Header chunk ntrks: 1
midifile: Header chunk division: 0x5A: 90 ticks per quarter note
midifile: Track chunk 0 type: MTrk, length 20
midifile: playing 1 track
note_on: 67 90
note_on: 64 90
note_on: 60 90
~~~

**Comment by dfkettle** [comment]
That's correct. As I said, I'm not getting the error messages in this patch (I'm still trying to come up with a simple example that does). The reason I posted it was to illustrate these other questions:

1. Although the message in the Pd console states that 43 bytes were written to the output file, if you open it in a text editor that can read binary files (Textpad or Notepad++ on Windows, I'm not sure what's available for Macs), you'll see that there are only 42 bytes in the file. At first, I thought this was of no consequence, but is it possible that under some conditions one byte which should have been written to the file wasn't written? Possibly a status byte, which would explain the error message about a missing status byte?

2. Sometimes (not always) I get this error message. I don't see it in your log file, so maybe it's a Windows-only issue.

`Unable to delete track file "E:/PureData/MIDI Sequencer/test.mid0.trk": Permission denied`

3. As you say, my patch plays back without error. But should it? Maybe this is just a misunderstanding on my part, but if I send a [track 0( message when recording, but then send a [track 1( message when playing it back, shouldn't I hear nothing, and/or see an error message about a missing track? If I understand the help patch correctly, it should play all tracks if you send [track -1(. but if you request a specific track, it should only play that track. If this is an issue and not just some confusion on my part, I can open another issue about this, since I don't think it's related to my original question.

Thanks.

**Comment by danomatika** [comment]
This is a process. First I started by running the patch to confirm if it was working the same on my system. I wasn't able to go into the finer points last night.

I would suggest breaking out your multiple issues into a [TODO task list](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#task-lists) in the initial issue text.

**Comment by dfkettle** [comment]
I added another issue to the task list. I'm not sure whether you want me to just keep adding issues to this same ticket, or create new tickets. I'm afraid that it could get a little confusing or unmanageable if I put everything in the same ticket. But let me know what your preference is.

**Comment by dfkettle** [comment]
I've been trying to compile 'midifile.c' just to see if I could figure out on my own whether or not there are really any issues with it, or whether it's just my lack of understanding (apart from the error in the help file, which I'll create another ticket for). I've never compiled any externals myself, but I do have some experience with C/C++, so I thought I'd give it a try. Anyway, I went through the tutorial (https://github.com/pure-data/externals-howto) and successfully compiled the examples in it using MSYS2 and 'pd-lib-builder'. But when I try to compile 'midifile.c' (without making any changes to it), I get a lot of warning messages (see attached log file). I know that they're just warnings and can probably be ignored, but is it normal to get all these warnings when compiling 'midifile.c'? Or am I doing something wrong (maybe incorrect compile options, although I thought 'pd-lib-builder' took care of setting the correct options). Could you ask whoever now maintains the MrPeach external (if anyone) if these messages are normal? My version seems to work, at least as far as I can see after doing some minimal testing, but some of the warning messages seem a little concerning, especially the ones about array subscripts being out of bounds.

[errors.log](https://github.com/pd-externals/mrpeach/files/8505493/errors.log)
